### DISCUSSION TITLE
Add `trackProperties` to access `$TRACK`

### DISCUSSION THREAD

**Discussion by otonashixav** [solidjs/solid#1077] [Ideas]
Similar to `unwrap` in stores. If people are going to use `$TRACK` explicitly, this would help with typescript and with possibly null targets, e.g.
```ts
/**
 * Track the properties of an object which implements `$TRACK` to track all its properties. 
 * Examples include `solid-js/store`. Does not track the properties of nested objects. 
 * Will NOT work on objects which do not implement `$TRACK`. 
 * 
 * @param target object to track. 
 * @returns `target`.
 */
function trackProperties<T>(target: T): T {
  return target && (target[$TRACK as keyof T], target);
}
```

(Naming is up for debate, since it's bound to cause confusion)

**Comment by ryansolid** [comment]
I have already started working on the `next` branch. With the intention of just having $TRACK return the proxy itself which simplifies the implementation a bit. But yeah exporting a function is probably nicer in a similar to how people don't use $RAW.

Naming is the other consideration. This is pretty explicit so I think it works. We really need to think about how to organize and document these sorts of advanced things. Because it's really easy to see this as a solution to a problem when the developer is solving the wrong problem, to begin with.

**Reply by trusktr** [reply]
> Because it's really easy to see this as a solution to a problem when the developer is solving the wrong problem, to begin with.

The `trackProperties` function would be very useful. Some people will abuse it, some people won't, but I don't think we can really solve that problem, and devs need flexibility with certainty (i.e. not accessing private API).

In the worst case, they're just downgrading to standard Lit/React/Preact _update-the-whole-component_ type of code, so it's not the end of the day.

**Reply by ryansolid** [reply]
I've been thinking exposing it through like `deep` helper as this would be a more efficient way to do deep tracking since it creates one signal per object rather than per property. If `deep` allowed optional depth then a depth of `0` would be this `trackProperties`
